### ISSUE TITLE
Gradient path does not apply `monomer_chain_break_indices`, so the design stages optimise a fused chain

### ISSUE THREAD

**Issue by moritztng** [PacesaLab/BindCraft2#20]
`monomer_chain_break_indices` (`bindcraft/af2.py:185`) has one call site: `_predict_complex` at
`af2.py:308`. The gradient path, `predict_complex_arrays` inside `_compiled_sequence_gradients`
(`af2.py:338-343`), builds `asym_id`, `entity_id` and `seq_mask` with the same three helpers
`_predict_complex` uses, then passes `residue_index` straight into `alphafold_input_features`.

For a monomer model that is the whole of chain identity. Monomer relpos reads `offset` clipped to
+-32 (`af/alphafold/model/modules.py:1469-1485`, `max_relative_feature` 32 in `af/alphafold/model/config.py:226`) and `asym_id` never
reaches it. So with `model_1_ptm` on a two-chain complex, `predict` and `sequence_gradients` hand
AlphaFold2 different molecules: the gradient path sees the binder and the target as one continuous
peptide, which is exactly what the docstring on `monomer_chain_break_indices` says must not happen.

### The difference is that one call, bit for bit

Binder 32 + target 96 built from sequence, `model_1_ptm`, `num_recycle=1`, `dropout=False`,
identical sequence parameters on both sides, CPU, one leg per process:

| entry point | pTM | i_pTM | pLDDT |
|---|---|---|---|
| `predict` | 0.294875 | 0.141805 | 0.363348 |
| `sequence_gradients` | 0.352421 | **0.522476** | 0.408438 |
| `predict`, with `monomer_chain_break_indices` replaced by the identity | 0.352421 | **0.522476** | 0.408438 |

Rows 2 and 3 agree on all four metrics **to the last bit** (`0.35242122411727905`,
`0.5224756002426147`, `0.4084381478605792`, `0.44594516418874264`), in separate processes. The
reverse closes too: feeding `sequence_gradients` a target whose own numbering already carries the
junction step of 50 reproduces row 1 bit for bit.

Two controls, both bit-exact:

- **One chain.** The branch at `:308` cannot fire, and the two entry points agree — pTM identical,
  pLDDT 8e-9 apart, which is float32 reduction order.
- **Zero pairs in the window.** Renumber the target so no cross-chain pair is within +-32 and the
  two numberings give the same answer to the last bit. The carrier is the relpos window, not the
  numbering as such.

### Where it lands in a trajectory

- `run_gradient_design_stage` takes its predictions from `sequence_gradients`
  (`trajectory.py:131`), and `judge_stage` grades `screen`, `refine`, `anneal` and `harden` on
  those same objects (`:310`). For a single-target campaign nothing replaces them: the pooling that
  would call `predict` is gated on `len(prepared_states) > 1` (`:178-183`).
- `run_sequence_mutation_stage` scores every candidate with `predict` (`:162`), and `mutate` is
  graded on `predict` (`:272-275`). `final` inherits those predictions (`:321-323`).

So a single-target trajectory optimises against one numbering for four stages and against the other
at mutate, and four of its six gates read a different instrument from the last two.

### How big, across binder draws

Gap is the gradient path's numbering minus the predict path's, same complex, one row per arm:

| binder | target | target numbered from | draws | in-window pairs | in-window share | i_pTM gap (each draw) | median |
|---|---|---|---|---|---|---|---|
| 32 | 96 | 1 | 4 | 1552 | 50.5% | +0.2317, +0.3433, +0.3807, +0.4749 | +0.3620 |
| 32 | 96 | 4000 | 1 | 0 | 0.0% | +0.0000 | +0.0000 |
| 32 | 160 | 1 | 4 | 1552 | 30.3% | +0.2283, +0.2857, +0.3812, +0.4090 | +0.3335 |
| 32 | 160 | 4000 | 1 | 0 | 0.0% | +0.0000 | +0.0000 |
| 32 | 288 | 1 | 3 | 1552 | 16.8% | +0.3213, +0.3847, +0.4237 | +0.3847 |
| 96 | 128 | 1 | 2 | 5712 | 46.5% | +0.1281, +0.3202 | +0.2242 |

The damage sits almost entirely in i_pTM. The same pairs, median over draws:

| binder | target | i_pTM | pTM | pLDDT | binder pLDDT |
|---|---|---|---|---|---|
| 32 | 96 | +0.3620 | +0.0280 | +0.0326 | +0.1061 |
| 32 | 160 | +0.3335 | -0.0256 | -0.0164 | +0.0310 |
| 32 | 288 | +0.3847 | -0.0056 | -0.0046 | -0.0189 |
| 96 | 128 | +0.2242 | +0.0625 | +0.0165 | +0.0346 |

pTM and pLDDT move by several times less and change sign between arms. i_pTM is what the stage
filters reject on.

All 13 draws that have cross-chain pairs inside the +-32 window show a gap of at least +0.128, and
both arms engineered to have no pairs in the window are bit-identical. Beyond that binary result we
will not fit a law: the gap moves by more than 0.2 between binder draws inside a single arm, the
arm medians overlap heavily, and neither the in-window pair count nor its share orders them. Going
from a 32-residue binder to a 96-residue one multiplies the in-window count by 3.7 and does not
raise the gap. We predicted a clean dose-response before running this and did not get one, so the
size is a range and not a formula.

### Which way to fix it is yours

Two readings fit: the gradient path is missing the call, or `_predict_complex` adds it
deliberately. We have no view on which you intend.

One argument that might favour leaving it out does not survive: there is no differentiability cost.
`sequence_design_loss` is differentiated with `jax.value_and_grad(..., argnums=2)`
(`af2.py:378`), and argnum 2 is `sequences`; `residue_index` arrives in `state_templates`, argnum 3,
which is not differentiated. We also ran the gradient path on the chain-broken numbering directly:
it compiles and returns a finite design loss and finite sequence gradients.

Happy to send a patch once you say which side should move.

### Reproducer

Pure upstream BindCraft 2 on CPU, no GPU and no structure file — both chains are built from
sequence. It needs `PYTHONPATH` pointed at the checkout and `params_model_1_ptm.npz` from
`bindcraft fetch-weights`. `census` needs neither the model nor the parameters. On 32 CPU cores a
`predict` leg at 128 residues takes about a minute including compile and the `sequence_gradients`
leg about three.

Each leg must run in its own process: both entry points memoise their traced function
(`prediction_compile_cache`, `gradient_compile_cache`) and neither cache key mentions the residue
numbering, so a patch applied inside a process that has already traced returns the old answer
bit-for-bit and reads as a null result.

<details>
<summary><code>chainbreak.py</code></summary>

```python
"""Do BindCraft 2's two AlphaFold2 entry points see the same molecule?

Pure upstream BindCraft 2, CPU only. No structure file: both chains are built from sequence, so a
clean checkout plus the monomer AlphaFold2 parameters is everything this needs.

`bindcraft/af2.py:305-306` renumbers a multi-chain complex before a monomer model sees it. The
gradient path at `af2.py:337-345` does not. This script measures what that costs on the metrics
BindCraft 2's own filters read.

One leg per process, because both entry points memoise their traced function and a patch installed
after a trace cannot move the answer.

  python chainbreak.py census                       # arithmetic only, no model, no parameters
  python chainbreak.py run --leg predict  --binder 32 --target 96 --params DIR
  python chainbreak.py run --leg gradient --binder 32 --target 96 --params DIR
"""
import argparse, json, os, platform, subprocess, sys, time

os.environ.setdefault('JAX_PLATFORMS', 'cpu')

AMINO_ACIDS = 'ACDEFGHIKLMNPQRSTVWY'

def deterministic_sequence(length, salt):
    """A fixed pseudo-random sequence. Content is irrelevant here; only that every leg gets the same one."""
    state, letters = 1234567 + 7919 * salt, []
    for _ in range(length):
        state = (1103515245 * state + 12345) % (1 << 31)
        letters.append(AMINO_ACIDS[state % 20])
    return ''.join(letters)

def residue_indices(binder_length, target_length, target_start):
    binder = list(range(1, binder_length + 1))
    target = list(range(target_start, target_start + target_length))
    return binder, target

def window_census(binder_length, target_length, target_start, window=32):
    """Cross-chain pairs the monomer relpos window admits, from the inputs alone."""
    binder, target = residue_indices(binder_length, target_length, target_start)
    inside = sum(1 for i in binder for j in target if abs(j - i) <= window)
    at_zero = sum(1 for i in binder for j in target if j == i)
    total = binder_length * target_length
    return {'binder_length': binder_length, 'target_length': target_length, 'target_start': target_start,
            'junction_step': target[0] - binder[-1], 'cross_chain_pairs': total,
            'in_window_pairs': inside, 'in_window_share': inside / total,
            'pairs_at_offset_zero': at_zero}

def census(args):
    rows = []
    for binder_length, target_length, target_start in args.arms:
        row = window_census(binder_length, target_length, target_start)
        broken = window_census(binder_length, target_length, target_start)
        # after monomer_chain_break_indices the junction step is MONOMER_CHAIN_GAP + 1 = 50, so the
        # target is renumbered to start 50 past the binder's last residue
        row['after_chain_break'] = window_census(binder_length, target_length, binder_length + 50)
        del broken
        rows.append(row)
    print(json.dumps(rows, indent=2))
    return rows

def build_states(binder_length, target_length, target_start, chains, seed=0):
    import jax, jax.numpy as jnp
    from bindcraft.protein import Protein, ResidueFlags

    def designed(length, salt):
        protein = Protein.from_fasta('>A\n' + deterministic_sequence(length, salt))
        return protein.replace(flags=jnp.full((length,), int(ResidueFlags.DESIGN), dtype=jnp.uint8))

    if chains == 1:
        return {'complex': {'binder': designed(binder_length + target_length, seed)}}
    target = Protein.from_fasta('>B\n' + deterministic_sequence(target_length, 1))
    target = target.replace(residue_index=jnp.arange(target_start, target_start + target_length, dtype=jnp.int32))
    return {'complex': {'binder': designed(binder_length, seed), 'target': target}}

def scalar_metrics(prediction, binder_length):
    import numpy as np
    metrics = prediction.metrics
    plddt = np.asarray(metrics['plddt'], dtype=np.float64)
    out = {'ptm': float(metrics['ptm']), 'plddt': float(plddt.mean()),
           'binder_plddt': float(plddt[:binder_length].mean())}
    if 'iptm' in metrics:
        out['iptm'] = float(metrics['iptm'])
    return out

def run(args):
    import jax, jax.numpy as jnp
    import bindcraft.af2 as af2
    from bindcraft.af2 import AlphaFoldDesignModel
    from bindcraft.loss import DesignLoss, iptm_loss, plddt_loss

    if args.leg == 'predict_identity':
        af2.monomer_chain_break_indices = lambda chain_lengths, residue_index: residue_index

    states = build_states(args.binder, args.target, args.target_start, args.chains, args.seed)
    binder_length = args.binder + args.target if args.chains == 1 else args.binder
    model = AlphaFoldDesignModel(presets='model_1_ptm', data_dir=args.params, num_recycle=args.recycle, dropout=False)
    sequence_parameters = dict(softmax_weight=1.0, one_hot_weight=1.0, temperature=0.01, logit_scale=2.0)

    started = time.time()
    if args.leg == 'gradient':
        loss_function = iptm_loss if args.chains > 1 else plddt_loss
        losses = {'objective': DesignLoss(function=loss_function, weight=1.0, required_states=frozenset({'complex'}),
                                          interface_mask=None)}
        if args.chains == 1:
            losses = {'objective': DesignLoss(function=lambda s, p: plddt_loss(s, p, prediction_state='complex', chain='binder'),
                                              weight=1.0, required_states=frozenset({'complex'}), interface_mask=None)}
        predictions, _, design_loss = model.sequence_gradients(states, losses, **sequence_parameters)
        result = scalar_metrics(predictions['complex'], binder_length)
        result['design_loss'] = float(design_loss)
    else:
        predictions = model.predict(states, **sequence_parameters)
        result = scalar_metrics(predictions['complex'], binder_length)

    result.update({'leg': args.leg, 'binder': args.binder, 'target': args.target,
                   'target_start': args.target_start, 'chains': args.chains, 'recycle': args.recycle, 'seed': args.seed,
                   'seconds': round(time.time() - started, 2), 'host': platform.node(),
                   'commit': subprocess.run(['git', 'rev-parse', '--short', 'HEAD'], cwd=os.path.dirname(af2.__file__),
                                            capture_output=True, text=True).stdout.strip(),
                   'jax': jax.__version__, 'loadavg': os.getloadavg()[0],
                   'threads': os.environ.get('XLA_FLAGS', '')})
    print(json.dumps(result))
    if args.out:
        with open(args.out, 'w') as handle:
            json.dump(result, handle, indent=2)
    return result

def arm(text):
    binder_length, target_length, target_start = (int(value) for value in text.split(','))
    return binder_length, target_length, target_start

def main():
    parser = argparse.ArgumentParser(description=__doc__)
    sub = parser.add_subparsers(dest='command', required=True)
    census_parser = sub.add_parser('census')
    census_parser.add_argument('--arms', type=arm, nargs='+',
                               default=[(32, 96, 1), (32, 160, 1), (32, 288, 1), (64, 192, 1), (32, 96, 4000)])
    census_parser.set_defaults(function=census)
    run_parser = sub.add_parser('run')
    run_parser.add_argument('--leg', choices=('predict', 'predict_identity', 'gradient'), required=True)
    run_parser.add_argument('--binder', type=int, default=32)
    run_parser.add_argument('--target', type=int, default=96)
    run_parser.add_argument('--target-start', type=int, default=1)
    run_parser.add_argument('--chains', type=int, default=2, choices=(1, 2))
    run_parser.add_argument('--recycle', type=int, default=1)
    run_parser.add_argument('--seed', type=int, default=0, help='picks the binder sequence; the target is held fixed')
    run_parser.add_argument('--params', default=os.environ.get('BC2_PARAMS', ''))
    run_parser.add_argument('--out', default='')
    run_parser.set_defaults(function=run)
    args = parser.parse_args()
    args.function(args)

if __name__ == '__main__':
    main()
```

</details>

Measured at `7a2dfdb`, re-read at `301efdd` where both the single call site and the stage/filter
split are unchanged.

Not claiming this is the cause of #13 — that was closed by the `aa_bias` fix in #17, and we have
not tested this against the PDL1-VHH case. It is in the same area and still live at HEAD, so it
seemed worth reporting separately.
